crates/matrix-sdk-crypto/src/machine.rs

```diff
@@ -715,47 +715,43 @@ impl OlmMachine {
         &self,
         decrypted: &mut OlmDecryptionInfo,
     ) -> OlmResult<()> {
-        let event: ToDeviceEvents = match decrypted.event.deserialize_as() {
-            Ok(e) => e,
-            Err(e) => {
-                warn!(
-                    sender = decrypted.sender.as_str(),
-                    sender_key = decrypted.sender_key.to_base64(),
-                    error = ?e,
-                    "Decrypted to-device event failed to be deserialized correctly"
-                );
-
-                return Ok(());
-            }
-        };
-
         trace!(
-            sender = decrypted.sender.as_str(),
-            sender_key = decrypted.sender_key.to_base64(),
-            event_type = %event.event_type(),
+            sender = %decrypted.sender,
+            sender_key = %decrypted.result.sender_key,
+            event_type = decrypted.result.event.event_type(),
             "Received a decrypted to-device event"
         );
 
-        match event {
-            ToDeviceEvents::RoomKey(e) => {
-                let session =
-                    self.add_room_key(decrypted.sender_key, decrypted.signing_key, &e).await?;
+        match &decrypted.result.event {
+            AnyDecryptedOlmEvent::RoomKey(e) => {
+                let session = self.add_room_key(decrypted.result.sender_key, e).await?;
                 decrypted.inbound_group_session = session;
             }
-            ToDeviceEvents::ForwardedRoomKey(e) => {
+            AnyDecryptedOlmEvent::ForwardedRoomKey(e) => {
                 let session = self
                     .key_request_machine
-                    .receive_forwarded_room_key(decrypted.sender_key, &e)
+                    .receive_forwarded_room_key(decrypted.result.sender_key, e)
                     .await?;
                 decrypted.inbound_group_session = session;
             }
-            ToDeviceEvents::SecretSend(mut e) => {
-                self.key_request_machine.receive_secret_event(decrypted.sender_key, &mut e).await?;
-                decrypted.event = Raw::from_json(to_raw_value(&e)?)
+            AnyDecryptedOlmEvent::SecretSend(e) => {
+                let name = self
+                    .key_request_machine
+                    .receive_secret_event(decrypted.result.sender_key, e)
+                    .await?;
+
+                // Set the secret name so other consumers of the event know
+                // what this event is about.
+                if let Ok(ToDeviceEvents::SecretSend(mut e)) =
+                    decrypted.result.raw_event.deserialize_as()
+                {
+                    e.content.secret_name = name;
+                    decrypted.result.raw_event = Raw::from_json(to_raw_value(&e)?)
```

Comment:
```suggestion
                    decrypted.result.raw_event = Raw::from_json(to_raw_value(&e)?);
```

Could also write the rhs as `Raw::new(&e)?.cast()`.